src/Transition.js

```diff
@@ -2,284 +2,274 @@ import React from 'react';
-    }
+  componentDidMount() {
+    this.performNextTransition();
   }
 
-  componentDidUpdate() {
-    this.processChild();
-  }
+  componentWillReceiveProps(nextProps) {
+    const status = this.state.status;
+    if (nextProps.in) {
+      if (status === EXITING) {
+        this.performEnter(nextProps);
+      } else if (this.props.unmountOnExit) {
+        // Perform enter in performNextTransition.
+        if (status === UNMOUNTED) {
+          this.setState({status: EXITED});
+        }
+      } else if (status === EXITED) {
+        this.performEnter(nextProps);
+      }
 
-  componentWillMount() {
-    this._mounted = true;
+      // Otherwise we're already entering or entered.
+    } else {
+      if (status === ENTERING || status === ENTERED) {
+        this.performExit(nextProps);
+      }
 
-    if (!this.props.transitionAppear) {
-      this.needsTransition = false;
-      this.setState({ in: this.props.in });
+      // Otherwise we're already exited or exiting.
     }
   }
 
-  componentWillUnmount(){
-    this._mounted = false;
+  componentDidUpdate() {
+    this.performNextTransition();
   }
 
-  componentDidMount() {
-    if (this.props.transitionAppear) {
-      this.processChild();
-    }
+  componentWillUnmount() {
+    this.cancelNextCallback();
   }
 
-  processChild(){
-    let needsTransition = this.needsTransition;
-    let enter = this.props.in;
-
-    if (needsTransition) {
-      this.needsTransition = false;
-      this[enter ? 'performEnter' : 'performLeave']();
+  performNextTransition() {
+    if (this.state.status === EXITED) {
+      if (this.props.in) {
+        // Either because of transitionAppear or unmountOnExit.
+        this.performEnter(this.props);
+      } else if (this.props.unmountOnExit) {
+        this.setState({status: UNMOUNTED});
+      }
     }
   }
 
-  performEnter() {
-    let maybeNode = React.findDOMNode(this);
-
-    let enter = node => {
-      node = this.props.transitioningNode(node) || node;
-
-      this.props.onEnter(node);
+  performEnter(props) {
+    this.cancelNextCallback();
+    const node = React.findDOMNode(this);
 
-      this.safeSetState({ in: true, transitioning: true, needInitialRender: false }, ()=> {
+    // Not this.props, because we might be about to receive new props.
+    props.onEnter(node);
 
-        this.props.onEntering(node);
+    this.safeSetState({status: ENTERING}, () => {
+      this.props.onEntering(node);
 
-        ensureTransitionEnd(node, () => {
-          if ( this.state.in ){
-            this.safeSetState({
-              transitioning: false
-            }, () => this.props.onEntered(node));
-          }
-
-        }, this.props.duration);
+      this.onTransitionEnd(node, () => {
+        this.safeSetState({status: ENTERED}, () => {
+          this.props.onEntered(node);
+        });
       });
-    };
-
-    if (maybeNode) {
-      enter(maybeNode);
-    }
-    else if (this.props.unmountOnExit) {
-      this._ensureNode(enter);
```

Comment:
Yes - let me rewrite this to be more explicit in that it's separately handling the `unmountAtExit` and `transitionAppear` cases.